### ISSUE TITLE
`import/namespace` is very slow

### ISSUE THREAD

**Issue by aaronadamsCA** [import-js/eslint-plugin-import#2340]
In our medium-sized mixed JS/TS codebase, disabling `import/namespace` significantly reduced our lint times. I haven't found this mentioned by anyone else, so I wanted to share the finding, because it was very surprising.

Our prior configuration ran like this:

```
Rule                              | Time (ms) | Relative
:---------------------------------|----------:|--------:
import/namespace                  |  5191.204 |    61.9%
import/no-relative-packages       |   609.625 |     7.3%
import/named                      |   453.670 |     5.4%
import/no-extraneous-dependencies |   281.672 |     3.4%
import/extensions                 |   213.542 |     2.5%
```

I always assumed this was just the "first rule tax", but adding `"import/namespace": "off"` did this for us:

```
Rule                              | Time (ms) | Relative
:---------------------------------|----------:|--------:
import/named                      |  1439.184 |    31.7%
import/no-relative-packages       |   571.847 |    12.6%
import/no-duplicates              |   395.090 |     8.7%
import/extensions                 |   391.228 |     8.6%
import/no-extraneous-dependencies |   267.072 |     5.9%
```

Given we already use `"import/no-namespace": "error"`, this rule wasn't doing anything for us anyway. With one line we've managed to significantly improve IDE responsiveness.

**Comment by ljharb** [comment]
Indeed, it's not necessarily that that rule is slow - it's that whichever rule is the first one to build up an ExportsMap of your entire codebase will be slow. Not every rule uses it.

However, `named` does use the ExportsMap, so something strange must be going on with the namespace rule. I'd be happy to review a PR that improved its performance.

**Comment by the-ult** [comment]
Same in our project.
Seems to be really slow:

```
Rule                               | Time (ms) | Relative
:----------------------------------|----------:|--------:
import/no-named-as-default         |  2807.942 |    17.4%
import/default                     |  2494.571 |    15.5%
import/order                       |  2422.542 |    15.0%
import/no-named-as-default-member  |  2124.755 |    13.2%
rxjs/finnish                       |  1958.145 |    12.2%
import/export                      |  1137.757 |     7.1%
import/no-unresolved               |   522.366 |     3.2%
sonarjs/no-ignored-return          |   476.080 |     3.0%
@nrwl/nx/enforce-module-boundaries |   381.036 |     2.4%
unused-imports/no-unused-imports   |   222.579 |     1.4%
Rule | Time (ms) | Relative
:----|----------:|--------:
```

Rules:
```json
"overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "plugins": [
             "unused-imports", 
             "rxjs-angular"
       ],
      "settings": {
        "import/parsers": {
          "@typescript-eslint/parser": [".ts", ".tsx"]
        },
        "import/resolver": {
          "typescript": {
            "alwaysTryTypes": true,
            "project": "./tsconfig.*?.json"
          }
        }
      },
      "extends": [
        "plugin:@nrwl/nx/angular",
        "plugin:unicorn/recommended",
        "plugin:sonarjs/recommended",
        "plugin:rxjs/recommended",
        "plugin:import/recommended",
        "plugin:import/typescript",
        "plugin:storybook/recommended",
        "plugin:@nrwl/nx/typescript"
      ],
      "parserOptions": {
        "project": ["tsconfig.*?.json"]
      },
      "rules": {
 
        "import/newline-after-import": "error",
        "import/no-unused-modules": "error",
        "import/order": [
          "error",
          {
            "alphabetize": {
              "order": "asc",
              "caseInsensitive": true
            },
            "newlines-between": "always",
            "pathGroupsExcludedImportTypes": ["builtin"]
          }
        ],
}
```

**Comment by aaronadamsCA** [comment]
@the-ult Make sure to compare total time with vs. without the rule, so you can rule out the building of the export map as the cause. It would be good to have another solid data point here!

**Comment by epmatsw** [comment]
Another data point here:

without namespace:

```
Rule                           | Time (ms) | Relative
:------------------------------|----------:|--------:
prettier/prettier              | 38484.339 |    43.1%
import/named                   | 13464.919 |    15.1%
react/no-deprecated            |  4004.563 |     4.5%
react/no-direct-mutation-state |  3019.211 |     3.4%
import/no-restricted-paths     |  2797.501 |     3.1%
import/extensions              |  2255.116 |     2.5%
react/no-string-refs           |  1929.564 |     2.2%
react/require-render-return    |  1392.313 |     1.6%
react/display-name             |  1247.210 |     1.4%
import/no-duplicates           |   888.921 |     1.0%
```

with namespace:

```
import/namespace                    | 68662.137 |    47.8%
prettier/prettier                   | 37565.273 |    26.2%
react/no-deprecated                 |  3884.333 |     2.7%
react/no-direct-mutation-state      |  2940.704 |     2.0%
import/no-restricted-paths          |  2334.704 |     1.6%
react/no-string-refs                |  1845.670 |     1.3%
import/extensions                   |  1786.525 |     1.2%
react/no-unstable-nested-components |  1482.991 |     1.0%
react/require-render-return         |  1371.941 |     1.0%
import/named                        |  1118.946 |     0.8%
```

eslint-plugin-import@2.25.4, eslint@7.32.0, node@14.18.0

**Comment by epmatsw** [comment]
Throwing some timing blocks around the visitors, it spends ~40% of the time in Program and ~60% in MemberExpression.

**Comment by ljharb** [comment]
I've pushed up b0e6f7f48945e2533e96d513248bf3e54e0c1aac but I don't expect it'll have a major impact on performance.

**Comment by epmatsw** [comment]
A somewhat informed guess: it looks like it's maybe cache missing on different relative paths to the same file: ex. `../../thing` vs `../thing`. I wonder if you could normalize those somehow with something like:

```
let importPath = declaration.source.value;
if (/\.+\//.test(importPath)) {
    importPath = path.resolve(path.dirname(context.getPhysicalFilename()), declaration.source.value);
}
```

**Comment by ljharb** [comment]
@epmatsw that does sound promising - if you wanted to try that locally (running the tests, and benchmarking it against your codebase) that would be very helpful :-) (even better, follow that up with a PR ;-) )

**Comment by epmatsw** [comment]
Seems like no dice, unfortunately. No apparent improvement from putting that snippet into Program :(

**Comment by digeomel** [comment]
I'm getting similar results on an angular-eslint project:

'''
Rule                                     | Time (ms) | Relative
:----------------------------------------|----------:|--------:
import/namespace                         |  4511.714 |    97.9%
import/no-unresolved                     |    70.773 |     1.5%
import/no-duplicates                     |    12.654 |     0.3%
@angular-eslint/component-selector       |     3.337 |     0.1%
import/export                            |     2.203 |     0.0%
import/no-named-as-default-member        |     1.985 |     0.0%
@angular-eslint/no-conflicting-lifecycle |     1.429 |     0.0%
@angular-eslint/no-output-rename         |     0.983 |     0.0%
import/default                           |     0.699 |     0.0%
@angular-eslint/contextual-lifecycle     |     0.690 |     0.0%
'''

As you can see, the top 3 rules are from this plugin.

**Comment by aaronadamsCA** [comment]
@digeomel Please make sure to compare **total run time** with the rule enabled vs. disabled.

I think it's now pretty well established that the rule has an internal slowdown, but if you want to add your data point, you'll need to provide that comparison.

**Comment by the-ult** [comment]
With import/order enabled in `eslint.json`
20sec
![Screenshot 2022-04-21 at 14 56 13](https://user-images.githubusercontent.com/4863062/164463931-9f6ce464-cb4a-49f7-8940-e78d8af9b6ca.png)

without:
13sec
![Screenshot 2022-04-21 at 15 02 40](https://user-images.githubusercontent.com/4863062/164463933-2051a3eb-329f-4fa0-b533-ff279ce7a650.png)

For a simple (small library)

**Comment by ljharb** [comment]
The `import/extensions` setting *always* needs to be set, so it's a bug in your config if it's not.

With that fixed config, what's the speed difference (with a bunch of import rules enabled) between "import/namespace on" and "import/namespace off"? (the eslint text output, please; screenshots of text aren't accessible)

**Comment by OutdatedVersion** [comment]
> The import/extensions setting always needs to be set, so it's a bug in your config if it's not.

In this case, the README could be clarified. Its current wording suggests it is optional, I believe:

> You may set the following settings in your .eslintrc:
> This defaults to ['.js']

Would it make sense to say "Despite the default, it is recommended to explicitly set `import/extensions` for your use-case."?

**Comment by ljharb** [comment]
Yes, that'd be great. It especially must be set if you're using "not standard javascript", namely, TS.

**Comment by the-ult** [comment]
> The import/extensions setting always needs to be set, so it's a bug in your config if it's not.

This is/was indeed not very clear in the documentation.

Could/should it be part of the ` - plugin:import/typescript` ??
Or be part of the documentation here?
- https://github.com/import-js/eslint-plugin-import#typescript

> With that fixed config, what's the speed difference (with a bunch of import rules enabled) between "import/namespace on" and "import/namespace off"? (the eslint text output, please; screenshots of text aren't accessible)

Tried to show it with the screenshots with the different build times.
But it seems like in some cases from `35sec` to `12sec` or even `1min` to `10sec`. So a major difference.

Will test/try it some more today

**Comment by JounQin** [comment]
In my business case, `import/no-duplicates` is very slow (I disabled `import/namespace` for `.ts` files).

**Comment by ljharb** [comment]
See https://github.com/import-js/eslint-plugin-import/issues/2340#issuecomment-1002398962; whichever is the first rule to build the ExportMap will be the slowest.

**Comment by tylerlaprade** [comment]
I have another data point from my own codebase. Of note is that my 2nd slowest `import/` rule (`no-unused-modules`) saw a minor increase in time, but nowhere close to the time saved by excluding `import/namespace`.

#### Rule On:
Total: **22,772 ms**
<img width="470" alt="image" src="https://github.com/import-js/eslint-plugin-import/assets/5475199/43f2ebf0-e5df-480d-a17e-abdf5c055654">

#### Rule Off:
Total: **16,133 ms**
<img width="472" alt="image" src="https://github.com/import-js/eslint-plugin-import/assets/5475199/bf7e6646-e80c-4d80-9967-b272c931ab98">

**Comment by jorisw** [comment]
Seeing as [apparently](https://github.com/import-js/eslint-plugin-import/issues/2340#issuecomment-1002398962) the first rule that builds the export map is always reported as the slowest, thus is misrepresenting time spent to the user, could this be seen as a bug in the reporting of the `TIMING` feature? This way we can't really know where to look for the bottleneck in our linter setup.

**Comment by ljharb** [comment]
Unfortunately since eslint doesn't provide any kind of "rule prep" stage, there's no way to separate out the export map timing from the rule timing.

**Comment by jorisw** [comment]
Could I put a meaningless rule in front that catches this prep timing, so the rest still is valid by proportion?

**Comment by ljharb** [comment]
hmm, maybe! I have no idea how eslint decides to order the rules.

What's probably easiest is to run timings; then disable the slowest rule and run them again, and in theory all but two should be roughly the same - and the one that's massively bigger in each run would be the one containing the export map buildup.

**Comment by soryy708** [comment]
Another performance suspect:

For each `VariableDeclarator` the rule visits, it performs DFS on child namespaces, which will have a larger impact if there are many such declarators, or if the dependency graph is large.
